packages/protocol/contracts/L1/libs/LibProposing.sol

```diff
@@ -45,47 +51,98 @@ library LibProposing {
         TaikoData.Config memory config,
         AddressResolver resolver,
         TaikoData.BlockMetadataInput memory input,
+        TaikoData.ProverAssignment memory assignment,
         bytes calldata txList
     )
         internal
         returns (TaikoData.BlockMetadata memory meta)
     {
+        // Check proposer
+        address proposer = resolver.resolve("proposer", true);
+        if (proposer != address(0) && msg.sender != proposer) {
+            revert L1_INVALID_PROPOSER();
+        }
+
+        // Check prover assignment
+        if (
+            assignment.prover == address(0) || assignment.prover == address(1)
+                || assignment.expiry <= block.timestamp
+        ) {
+            revert L1_INVALID_ASSIGNMENT();
+        }
+
+        // Too many unverified blocks?
+        TaikoData.SlotB memory b = state.slotB;
+        if (b.numBlocks >= b.lastVerifiedBlockId + config.blockMaxProposals + 1)
         {
-            address proposer = resolver.resolve("proposer", true);
-            if (proposer != address(0) && msg.sender != proposer) {
-                revert L1_PERMISSION_DENIED();
-            }
+            revert L1_TOO_MANY_BLOCKS();
         }
-        // Try to select a prover first to revert as earlier as possible
-        (address assignedProver, uint32 rewardPerGas) = IProverPool(
-            resolver.resolve("prover_pool", false)
-        ).assignProver(state.slotB.numBlocks, state.slotC.feePerGas);
 
-        assert(assignedProver != address(1));
+        if (config.skipProverAssignmentVerificaiton) {
+            // For testing only
+            assignment.prover.sendEther(msg.value);
+        } else {
+            // Verify prover assignment and pay the prover Ether as proving fee.
+            // Note that this payment is permanent. If the prover failed to
+            // prove the block, its bond is used to pay the actual prover.
+            if (assignment.prover.isContract()) {
+                IProver(assignment.prover).onBlockAssigned{ value: msg.value }(
+                    input, assignment
+                );
+
+                // Burn the prover's bond
+                TaikoToken(resolver.resolve("taiko_token", false)).burn(
+                    assignment.prover, config.proofBond
```

Comment:
The original idea is to allow Prover contract to manage it's own Taiko token asset.
I change the code to burn from `taikoTokenBalances` if there are enough tokens, otherwise, burn from the prover address, for both contracts and EOAs.